src/wasm/wasm-validator.cpp

```diff
@@ -2213,19 +2213,31 @@ void FunctionValidator::visitRefCast(RefCast* curr) {
     getModule()->features.hasGC(), curr, "ref.cast requires gc to be enabled");
   if (curr->ref->type != Type::unreachable) {
     shouldBeTrue(
-      curr->ref->type.isRef(), curr, "ref.test ref must have ref type");
+      curr->ref->type.isRef(), curr, "ref.cast ref must have ref type");
   }
   if (curr->rtt->type != Type::unreachable) {
     shouldBeTrue(
-      curr->rtt->type.isRtt(), curr, "ref.test rtt must have rtt type");
+      curr->rtt->type.isRtt(), curr, "ref.cast rtt must have rtt type");
   }
 }
 
 void FunctionValidator::visitBrOnCast(BrOnCast* curr) {
   shouldBeTrue(getModule()->features.hasGC(),
                curr,
                "br_on_cast requires gc to be enabled");
-  WASM_UNREACHABLE("TODO (gc): br_on_cast");
+  if (curr->ref->type != Type::unreachable) {
+    shouldBeTrue(
+      curr->ref->type.isRef(), curr, "br_on_cast ref must have ref type");
+  }
+  if (curr->rtt->type != Type::unreachable) {
+    shouldBeTrue(
+      curr->rtt->type.isRtt(), curr, "br_on_cast rtt must have rtt type");
+    shouldBeEqual(curr->rtt->type.getHeapType(),
+                  curr->castType.getHeapType(),
+                  curr,
+                  "br_on_cast rtt must have the proper heap type");
+    noteBreak(curr->name, Type(curr->rtt->type.getHeapType(), Nullable), curr);
```

Comment:
```
noteBreak(curr->name, curr->castType, curr);
```

Reply:
Thanks! Followup in https://github.com/WebAssembly/binaryen/pull/3457